### ISSUE TITLE
Weather system

### ISSUE THREAD

**Issue by AgentRev** [A3Armory/A3A_A3Wasteland_Sock.Altis#33]
I would like to understand what are the motivations behind replacing the vanilla weather script with your current one.

**Comment by codexGW** [comment]
Had a couple bugs that were bothering me and I tried to fix them but I was having a hard time understanding what exactly some of the script was doing. It would rain for only a few seconds with not a cloud in the sky and there was this weird rolling fog that would come in back and forth randomly and stop about 500 meters away. If you check out the history for the old script you can see a lot of the attempts. Once the weather saving was implemented I looked for simpler script and came across the current script. I just reorganized and changed a few things and converted it to English. I really like the simplicity of the new script and the low elevation fog was a cool bonus. Great way to get cover for approaching a mountaintop mission or bases.

**Comment by AgentRev** [comment]
I'm gonna be blunt, your current system is really crappy. It starts raining full-on for 60 seconds with absolutely no clouds or fog, then it stop right off with full fog and a rainbow. Just a couple hours ago it went on full pea soup fog with no visibility for 10 minutes then it disappeared instantly.

Fog decay syncing is messed-up engine-side; I [reported](http://webcache.googleusercontent.com/search?q=cache:feedback.arma3.com/bug_revision_view_page.php%3Fbug_id%3D22415) the bug a year ago but BIS can't be bothered to read it like usual. My script tries to apply the correct server values, but native syncing tries to overwrite it every 3 seconds with incorrect values. I did however forget to revert to the classic fog method, which I will do now that you reminded me.

I believe the rain that comes on and off with the vanilla DWS is actually the engine trying to trigger rain but the script overrides it off. This only happens at mission start. The problem here would be the engine, not the script.

Simplicity != good. People complain to me how A3W is complicated all the time, but the complexity is not in vain. Every little detail is taken care of. In the vanilla DWS, rain cannot start until overcast reaches 75%, because rain without clouds is illogical; for every increase in rain, there is an associated 1/4 increase in fog, as to simulate the reduction in visibility which is otherwise not taken care of; weather changes are spread out on a bigger timescale, as to reflect real-life weather behavior.

**Comment by codexGW** [comment]
I didn't make this [script](http://www.altisliferpg.com/topic/5699-dynamicrandom-weather/) and people seemed to like it and not have issues with it. I guess something got screwed up or maybe its not the script at all because it looks fine and works as the BIS says it should. Maybe it's the game and info provided that's crap. From what I understand the [60](https://github.com/A3Armory/A3A_A3Wasteland_Sock.Altis/blob/dev/addons/scripts/SimpleWeatherScript.sqf#L101) is the time before it starts and from what I [read](https://community.bistudio.com/wiki/setOvercast) it's not supposed to rain unless setOvercast is .7 or over. I guess I'll try and set all the rain chances to setOvercast .7 or higher and set the time in front of setRain to 60+

**Comment by AgentRev** [comment]
What I mean is that there are all sorts of little issues like that during the course of the mission. All your script does is roll a dice and pick values from hardcoded lists, causing nonsensical weather changes from 0 to 100 way too fast.

On the other hand, the DWS modulates the values over time, slightly notching them up or down every couple seconds, without causing radical transitions. Have you ever played during an hour-long fluctuating thunderstorm with high overcast? It's definitely enjoyable.

> it's not supposed to rain unless setOvercast is .7 or over

Rookie mistake. A setOvercast change from 0 to 1 takes at least a full 90 minutes of real time. If you set overcast to 1 and rain to 1, it's actually gonna start raining once overcast reaches 0.5, which has almost no clouds. Overcast from 0.5 to 0.75 takes about 30 minutes long. The bad weather syncing is making this worse, since overcast is not correctly forced, and 0.5 can appear as a slightly-dark sky with no clouds and a dimmed sun, like smog or some shit. You end up with [this](http://i.imgur.com/3PC34J0.png).

I can and will fix the issues, but you should definitely revert to the DWS, your SWS completely breaks immersion at least once every evening I play.

**Comment by codexGW** [comment]
Wow, weather is all fucked up and I'll definitely defer to your knowledge about it. I don't know why they just don't sync whats going on server side with the client. I don't see the benefit of having separate values. I'll revert to DWS. Will there be any issues with not having weather persistence?

Please consider making it easier to set chances for bad weather. Also if you can implement more low ground fog as a choice. It's really cool looking especially at sunset and sunrise.
http://i.imgur.com/jQ63K7ol.jpg
http://i.imgur.com/pJxPP1ml.png
Oh another thing, the script http://i.imgur.com/V3UdU6X.png

**Comment by AgentRev** [comment]
![cowbell](https://media0.giphy.com/media/whOs1JywNpe6c/200.gif)

**Comment by AgentRev** [comment]
I've done some minor changes to the weather system. The rain issue should not be a problem anymore, I found a way to enforce it correctly on mission start.

I am not able to implement fog decay and height variation, because this shit is fubar to the bone. If you set a high fog decay and a height over 10 to 20m, everything at sea level is engulfed into the thickest fog the game is able to produce, regardless of which fog value you actually set, as if the fog decay is applied in reverse too (constant increase in density below base height). I suspect this is what caused the pea soup issue I talked about. It might look good from the hills, but people at sea level can't see jackshit and that's a big problem IMO. Also, I need to keep the fog decay at 0 to have correct rain visibility at all heights, and the value imprecision bug I referenced earlier makes decay transition look wonky af.

**Comment by codexGW** [comment]
Right on man. Thanks for testing this. I'll keep all the settings default for awhile and I'll make sure to check it out.

**Comment by codexGW** [comment]
I was kind of sleep walking when I read your comment last time and did that merge. That is a bummer about the low ground fog. I don't mind the pea soup so much as long as its low enough and super rare. Maybe you can make it only waist level. I like how it flushes the campers out for a bit and totally changes game-play and where the action is. It's cool I think if that only happens a couple times a week. Either way its not that big of a deal to me or the server I think. I really do like spooky Arma though :smile:

**Comment by mooneymoon** [comment]
Wind is way too strong since the merge. Smoke canisters will bounce a few inches into the air and fly 20m.

**Comment by AgentRev** [comment]
Max wind currently approximates to 31.3 mph, adjust the _maxWind value to your liking

**Comment by codexGW** [comment]
Lets leave it default for now. Just like real life the weather doesn't always cooperate. This isn't happening all the time or way to much, right?

**Comment by mooneymoon** [comment]
I played for about 2-3 hours earlier and it was too strong the entire time. I'll pay more attention when I play later.

But yeah, too much.

**Comment by AgentRev** [comment]
I am noticing there are still issues with fog... looks like I need to check that again

**Comment by codexGW** [comment]
If you're talking about on our servers, I haven't updated them with your latest commit yet. The servers update usually after I merge the other repos. I'm going to update them tonight. I've been trying to figure out what's going on with the maxping not working.

**Comment by AgentRev** [comment]
Oh yeah, my comment was about your Stratis server.

**Comment by codexGW** [comment]
Servers have been updated to the latest default DWS.

**Comment by AgentRev** [comment]
Okay, fog is definitely more stable now. I noticed there are some issues with overcast syncing, which caused a bit of rain to appear in sunlight. I think it may have to do with timeMultiplier acceleration not speeding up visual changes correctly. I don't know if this can be solved without having to slow down the weather script.

**Comment by AgentRev** [comment]
Yep I just tested, weather transition time is sped up according to the timeMultiplier, so a potential solution would be to divide transition time by the multiplier, I'll check if that works.

**Comment by codexGW** [comment]
That explains why it was so bad during the night cycle. The timeMultiplier is set to 4 then. I'll update the servers with your fix soon.

**Comment by codexGW** [comment]
We still have reports of wind gust blowing parachutes and smoke grande off course. Is there a max wind gust? I don't think the max of 6 could be doing this. I also saw it rain strangely yesterday on Stratis. I was in game for about 20 minutes when it started raining. There was very little cloud cover and after about 15 minutes of raining somewhat hard there still wasn't more then a few clouds and none overhead. It did get a bit more hazy/foggy.

**Comment by AgentRev** [comment]
The DWS does not interact with `setGusts`. I don't know how those work behind the scene, I've never investigated gusts at all.

If you relog you will see that the clouds refresh to the correct value... my experiments indicate cloud transition is piss-poor and it's not specific to the DWS. Cloud clover desyncs easily even in the editor. If you call `simulWeatherSync` when having a hazy horizon, you will see the clouds sync to the proper value. This command cannot be used in the DWS because it completely freezes the game for a couple seconds.

**Comment by codexGW** [comment]
I'll just leave it as is. The settings are so fucked up and overly complicated for changing weather dynamically this is probably the best it can get. Thanks for the info Agent.

**Comment by AgentRev** [comment]
I cannot reproduce wind issues no matter how hard I try. If you could get a recording, that might help, because I've never seen it in person. Maybe also grab the value of `wind`.

Also I tried disabling DWS syncing to see if native syncing was any better, but it's definitely worse.

**Comment by codexGW** [comment]
I haven't seen the strong wind gust either. I have only been in game probably only a hour the past month though. I don't really have a problem with it myself if its rare. I think the weather shouldn't always cooperate. I will ask mooney and anyone that reports it again for a video and definitely keep an eye out for it myself. Thank you Agent.

**Comment by mooneymoon** [comment]
Whatever you guys did in the last week or two seems to have fixed it.

**Comment by codexGW** [comment]
Agent did the fixing. I think the person that last it reported was saving issues to post and didn't realize it was fixed. When I asked some other people more about it, all I was told is that it was not as bad, not fixed or resolved. Sorry Agent for any confusion about the wind gust.

**Comment by mooneymoon** [comment]
2 things are out of whack still.

The wind is weaker, but still too strong too often.

There is fog too often.

These are supposed to be Mediterranean islands so I would expect sun most of the time. I'm not from the coast but shouldn't the ocean breeze also make fog pretty much impossible?

Anyway there is my $0.02.

**Comment by codexGW** [comment]
Isn't it funny we feel our thoughts are worth 2 cents but only a penny when we want someone else's. I read that on shower thoughts yesterday :smile: 

This should fix the haziness
https://github.com/A3Armory/A3A_A3Wasteland_Sock.Altis/commit/4877ed9c071ebe8c96eda58e09d5d87af6f8f528

**Comment by codexGW** [comment]
The strangest thing just happened to me on A3A Stratis. I logged in and it was fine but I was testing the team unlock so I re-logged and the weather was completely different. It was raining, full overcast, and even some lightning. When I asked the only person that was on if it was raining he said no and the weather was sunny. I even spectated him to be sure. Its been 20 minutes and its still completely overcast. Stopped raining though. I even re-logged again, hoping that would fix it. I think something is off with the sync. This is probably why we get all these reports and forum post about it but they are inconsistent.

http://www.a3armory.com/forum/thread-590-post-4098.html#pid4098
http://www.a3armory.com/forum/thread-551-post-4073.html#pid4073

**Comment by AgentRev** [comment]
Yes, overcast is really messed up right now... For example, the server will not attempt to trigger rain until `overcast` reaches 0.75, however on the client it can start raining full-power once it reaches exactly 0.5, which is quite abnormal.

I did some changes to weather syncing in my last couple edits; previously, the weather was being force-broadcast to everyone every time a player joined the server, which doesn't make much sense. Instead, I changed it so it will only force-broadcast to the new player. There could be a connection, but I'm not 100% sure. Maybe it was the last barrier against the engine itself broadcasting bullshit values.

**Comment by codexGW** [comment]
I just want to add I didn't even think to restart Arma. I was kinda of in zombie mode at that point. Way past my bedtime. It was weird though for 30 minutes plus I was stuck in full overcast (.7+). I reconnected to the server 3 times too. I kept asking the player if it was overcast for him and he would say it was mostly sunny.

**Comment by codexGW** [comment]
This shit makes me really think twice about supporting suck a wonky game with crap support. Such a waste of resources, especially with you. You could have done the whole kill message thing or something else with all the effort you spent on this. Man sometimes I got to say, fuck BIS.

**Comment by AgentRev** [comment]
I'd be beyond the sock migration if I didn't have to constantly make workarounds for shit that doesn't work like it's supposed to every couple months.

**Comment by AgentRev** [comment]
On the weather problem, I remember me and fellow BP dev Deathlyrage asked BIS for an option to disable weather syncing because it's total shit, Dwarden replied "disabling something instead of fixing it isn't exactly an option" Thus, nothing else was ever done about it since then.

ALSO, looks like a new feedback tracker is online: https://feedback.bistudio.com/
It's about fucking time.

Here is the weather report: https://feedback.bistudio.com/T83052

**Comment by codexGW** [comment]
That just pissed me off reading that tracker. I think I will cool off a bit before adding too it. The arrogance in that one comment is very disappointing. Especially with that much talent commenting. This is what I saw "We change (break) shit and if you can't fix it with your limited time and resources, fuck you and fuck the people playing your mods".

Should we just revert to the commit before the last one? The syncing seems to be worse and seems to favor bad weather.

**Comment by codexGW** [comment]
Imagine where Arma would be if they put the resources they put towards one DLC to supporting their top community devs as much as possible for a year. I'm not talking a contest which fractures open community's either. Just a year of listening and working with the top mods and this game would make so much more money. Like you said before, where would this game be without developers like Tonic? BIS need to starting thinking about the long game before they start losing more community development. End of rant.

**Comment by AgentRev** [comment]
Ironically, lack of support is pretty much the main reason why Tonic quit Arma.

**Comment by AgentRev** [comment]
They closed the tracker again. Looks like they couldn't handle providing support for the first time in 3 months.

**Comment by codexGW** [comment]
I swear they think its just a fps and not a military sim with community development. What do they think keeps it going? It certainly isn't all the bugs and under 60 frames per second performance.

**Comment by AgentRev** [comment]
Completely unrelated to weather, but I just discovered that players wearing the default OPFOR uniform are 6 times tougher than those wearing the default BLUFOR uniform... You can one-hit-kill BLUFOR in the chest with one .45 shot, but OPFOR will need at least 4 shots... What the fuck... Stupid shit game... Fucking BIS... Do I really have to make a goddamn workaround again? Geez...

**Comment by mooneymoon** [comment]
![heh](http://i.imgur.com/zSXpCfW.png)

**Comment by AgentRev** [comment]
There is a logical reason behind that; the CSAT faction does not have any ballistic vest... Most CSAT units only wear an LBV Harness or Slash Bandolier, which have no armor, while the heavier units wear lower-end Tactical Vests. All the big vests are NATO or AAF specific. So BIS probably though "hurr durr too lazy to make CSAT vests, lets give them armor by default!!!1!"

Gross lack of judgment.

**Comment by AgentRev** [comment]
There is some sort of workaround... "Officer Fatigues (CSAT)" (`U_O_OfficerUniform_ocamo`) does not appear to have extra armor... So I guess we could give fresh spawns that one instead.

However, all OPFOR ghillie suits have extra armor too, and I'm not sure if I can do much for that.
What a shitfest.

**Comment by codexGW** [comment]
Wow. This explains a lot. I see why op4 is so popular. I wonder if most people know this or its just subconscious.

**Comment by AgentRev** [comment]
Personally I go Opfor because I wanna play with a team and all the noobs are always Blufor.

**Comment by AgentRev** [comment]
Also the NATO MTP camo stands out like a sore thumb in the sun.

**Comment by mooneymoon** [comment]
Can one of you ELI5 why _forceAddUniform_ isn't just used throughout the mission? I'm not necessarily saying it should be I'm just wondering.

**Comment by AgentRev** [comment]
If you are wearing a prohibited uniform and you drop it on the ground or put it in a crate, you can't pick it up. That's pretty much all there is to it. I've always dearly hated this shit.

**Comment by AgentRev** [comment]
BIS are all like "hurr durr we must respect the United Nations resolution that says it is illegal to wear an enemy uniform". That kinda faget shit.

**Comment by AgentRev** [comment]
I suggest we continue in this new issue I just made.

**Comment by codexGW** [comment]
Close to a epic issue. lol

**Comment by AgentRev** [comment]
I think I'm gonna call `simulWeatherSync` each time the respawn menu opens... this should make cloud cover more consistent for all players

**Comment by codexGW** [comment]
Cool, I'm down to test it. If you want we can just throw it on one server too.

**Comment by codexGW** [comment]
I was thinking a few days ago about redoing the simple weather script and just use the setovercast setting and use the 4 arrays of weather options to be a progression of good to bad with a setvariable so it will only use the next progression. I figured maybe if you let the engine do its thing with the rest of the settings it might be synced. This probably would not have worked though. :smile:

**Comment by AgentRev** [comment]
> if you let the engine do its thing

oh so you mean, nothing at all?

**Comment by codexGW** [comment]
What I mean is, I have seen it rain and then have thunder/lightning if setovercast set high enough. It's just wishful thinking that the rest would follow and all would be synced lol

**Comment by AgentRev** [comment]
Thunder will happen whenever overcast is high enough, it's not managed by the DWS. Rain is not the problem in itself, it's because the DWS is not very well coded overall and it can't keep track of fog separately from the current readings. The fog during rain is something I implemented myself, it's not part of the game or the original DWS... I do not want rain without fog, because otherwise you still have perfect visibility when it's raining full power, which makes no sense. The DWS does not keep track of fog correctly, so rain is out of the question.

**Comment by codexGW** [comment]
I'm going to miss those rainbows.

**Comment by AgentRev** [comment]
Even with all of this, overcast syncing is still completely fubar engine-side. Rain tries to be constant server-wide, but not all players have the same overcast level, and cloud cover desyncs away really easily from overcast, so you have rain in the sun.

**Comment by codexGW** [comment]
Yeah I really hate that. It totally kills the immersion.
